templates/commands/render/templatesource/source.go

```diff
@@ -0,0 +1,99 @@
+	// makes the linter stop complaining.
+	_ sourceParser
+	_ templateDownloader
+	_ = realSourceParsers
+	_ = parseSource
+)
+
+// sourceParser is implemented for each particular kind of template source (git,
+// local file, etc.).
+type sourceParser interface {
+	// sourceParse attempts to parse the given src. If the src is recognized as
+	// being downloadable by this sourceParser, then it returns true, along with
+	// a downloader that can download that template.
+	sourceParse(ctx context.Context, src, protocol string) (templateDownloader, bool, error)
+}
+
+// A templateDownloader is returned by a sourceParser, and offers the ability to
+// download a template.
+type templateDownloader interface {
+	// Download downloads this template into the given directory.
+	Download(ctx context.Context, outDir string) error
+}
+
+// realSourceParsers contains the non-test sourceParsers.
+var realSourceParsers = []sourceParser{
+	// This source parser recognizes template sources like
+	// "github.com/myorg/myrepo@v1.2.3" (and variants thereof).
+	&gitSourceParser{
+		re: regexp.MustCompile(
+			`^` + // Anchor the start, must match the entire input
+				`(?P<host>github\.com|gitlab\.com)` + // The domain names of known git hosting services
+				`/` +
+				`(?P<org>[a-zA-Z0-9_-]+)` + // the github org name, e.g. "abcxyz"
+				`/` +
+				`(?P<repo>[a-zA-Z0-9_-]+)` + // the github repo name, e.g. "abc"
+				`(/(?P<subdir>[^@]*))?` + // Optional subdir with leading slash; the leading slash is not part of capturing group ${subdir}
+				`@(?P<version>[a-zA-Z0-9_/.-]+)` + // The "@latest" or "@v1.2.3" or "@v1.2.3-foo" at the end; the "@" is not part of the capturing group
+				`$`), // Anchor the end, must match the entire input
+		httpsRemoteExpansion: `https://${host}/${org}/${repo}.git`,
+		sshRemoteExpansion:   `git@${host}:${org}/${repo}.git`,
+		subdirExpansion:      `${subdir}`,
+		versionExpansion:     `${version}`,
+	},
+
+	// More sourceParsers are coming imminently for:
+	//  - local template directories
+	//  - go-getter-style git URLs
+}
+
+// parseSource maps the input template source to a particular kind of source
+// (e.g. git) and returns a downloader that will download that source.
+//
+// source is a template location, like "github.com/foo/bar@v1.2.3". protocol
+// is the value of the --protocol flag, like "https".
+//
+// A list of sourceParsers is accepted as input for the purpose of testing,
+// rather than hardcoding the real list of sourceParsers.
+func parseSource(ctx context.Context, srcParsers []sourceParser, source, protocol string) (templateDownloader, error) {
+	var dl templateDownloader
+	for _, sp := range srcParsers {
+		downloader, ok, err := sp.sourceParse(ctx, source, protocol)
+		if err != nil {
+			return nil, err //nolint:wrapcheck
+		}
+		if ok {
+			dl = downloader
+			break
+		}
+	}
+	if dl == nil {
+		return nil, fmt.Errorf("template source %q isn't something that we know how to download", source)
+	}
+
+	return dl, nil
```

Comment:
Could this be simplified to:

```suggestion
	for _, sp := range srcParsers {
		downloader, ok, err := sp.sourceParse(ctx, source, protocol)
		if err != nil {
			return nil, err //nolint:wrapcheck
		}
		if ok {
		  return downloader
		}
	}
	
	return nil, fmt.Errorf("template source %q isn't something that we know how to download", source)
```